### ISSUE TITLE
A shard gets an error of: "data_dictionary::no_such_keyspace" after a column-definitions schema changes

### ISSUE THREAD

**Issue by yarongilor** [scylladb/scylladb#12491]
## Issue description

- [ ] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

*Describe your issue in detail and steps it took to produce it.*

## Impact

*Describe the impact this issue causes to the user.*

## How frequently does it reproduce?

*Describe the frequency with how this issue can be reproduced.*

## Installation details

Kernel Version: 5.15.0-1026-aws
Scylla version (or git commit hash): `2022.1.5-20230109.8c2c21866` with build-id `f6a36531f7d856e02ab87c633ffae271aaf57396`

Cluster size: 4 nodes (i3.2xlarge)

Scylla Nodes used in this run:
- longevity-lwt-3h-2022-1-db-node-159ead0e-4 (3.219.167.246 | 10.0.3.223) (shards: 8)
- longevity-lwt-3h-2022-1-db-node-159ead0e-3 (3.238.105.23 | 10.0.0.188) (shards: 8)
- longevity-lwt-3h-2022-1-db-node-159ead0e-2 (18.213.113.134 | 10.0.2.247) (shards: 8)
- longevity-lwt-3h-2022-1-db-node-159ead0e-1 (3.236.134.206 | 10.0.1.233) (shards: 8)

OS / Image: `ami-0fdaa28349f8c730d` (aws: us-east-1)

Test: `longevity-lwt-3h-test`
Test id: `159ead0e-bb07-4cd7-aa0f-f90d9f1a015d`
Test name: `enterprise-2022.1/longevity/longevity-lwt-3h-test`
Test config file(s):

- [longevity-lwt-basic-3h.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/721d9a4c443292bad00e251991a5e86ea5b4aa6a/test-cases/longevity/longevity-lwt-basic-3h.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor 159ead0e-bb07-4cd7-aa0f-f90d9f1a015d`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=159ead0e-bb07-4cd7-aa0f-f90d9f1a015d)
- Show all stored logs command: `$ hydra investigate show-logs 159ead0e-bb07-4cd7-aa0f-f90d9f1a015d`

## Logs:
- **db-cluster-159ead0e.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/db-cluster-159ead0e.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/db-cluster-159ead0e.tar.gz)
- **monitor-set-159ead0e.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/monitor-set-159ead0e.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/monitor-set-159ead0e.tar.gz)
- **loader-set-159ead0e.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/loader-set-159ead0e.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/loader-set-159ead0e.tar.gz)
- **sct-runner-159ead0e.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/sct-runner-159ead0e.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/sct-runner-159ead0e.tar.gz)
- **parallel-timelines-report-159ead0e.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/parallel-timelines-report-159ead0e.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/159ead0e-bb07-4cd7-aa0f-f90d9f1a015d/20230110_010200/parallel-timelines-report-159ead0e.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/enterprise-2022.1/job/longevity/job/longevity-lwt-3h-test/31/)
</details>

**Comment by yarongilor** [comment]
A possibly similar error of "can't find a keyspace" in https://github.com/scylladb/scylladb/issues/7901 might be related to this error.

**Comment by mykaul** [comment]
@kostja - can someone from your team take a quick look (mainly looking to ensure it's not a regression)

**Comment by kostja** [comment]
@yarongilor is raft on or off?

**Comment by yarongilor** [comment]
The memory usage doesn't look clear, can try investigating it as well:
![image](https://user-images.githubusercontent.com/9427929/211787889-30bfbe45-6cb4-440b-a5bd-79141a360bb6.png)

**Comment by mykaul** [comment]
> @yarongilor is raft on or off?

Off, it's 2022.1.5 testing

**Comment by avikivity** [comment]
Very odd. no_such_keyspace after ALTER TABLE ADD COLUMN?

I think this should be reproducable locally with a single-node cluster, given the instructions above.

**Comment by kostja** [comment]
no_such_keyspace expects a keyspace name as a parameter which should be formatted into the error message, and in this log it's missing.

**Comment by kostja** [comment]
@avikivity exceptions like these should capture the backtrace. It would make it much easier to analyse the root cause.

**Comment by kostja** [comment]
So here's my guess for the root cause of the issue.

First of all, we're dealing with the same grand LWT correctness test designed by @juliayakovlev which uses views.

Second, the exception is printed from the storage proxy code handling the write on the coordiantor side, but happens on the replica side, when executing the write.

When executing the write we normally do not look into the keyspace any more, we take the schema object and apply the mutation. So why would we want to look up a keyspace at all, and on top of that use a corrupt keyspace name?

My thinking it's because the write applies to a view. In this case we will go into the migration manager to fetch the schema_ptr for the write, it will notice that the schema object is a view, and will try to look up the base table for it. 

For this look up, it will try to use the view info object, which, as we learned from https://github.com/scylladb/scylladb/issues/11542 can be corrupt after a DDL. 

This is how we get an empty keyspace name, we get it from corrupt view_info in      schema_ptr base_schema = db.find_schema(us->view_info()->base_id());         in get_schema_definition() call. 

So the question is, is the fix for #11542 included into the build? If yes, it may be another manifestation of the bug or a related problem.

Either way my quick guess it's not a regression, but given the keyspace name is empty (it should never be) it is a genuine bug. The next steps would be:
- make sure the fix for #11542 is included and re-run the test
- if the bug still is present, reduce the test case to one that can fit into test.py environment
- look for the root cause in the view code space.

**Comment by nyh** [comment]
@kostja this base-lookup code is extremely convoluted and was written as a patch over patch over patch, each time debugged and fixed by a different person, so it's almost impossible to understand exactly how it works. I don't think that #11986 would have fixed something that looks like this, but it's really hard to be sure. 

Anyway I looked at git log in 2022.1.5-20230109.8c2c21866 and that patch - "materialized views: fix view writes after base table schema change" - is already in it. So it's a different, although probably similar, bug.

I'll need to start thinking how to reproduce this in a sort of test that *I* can actually run. I see no other way to debug it myself.

**Comment by kostja** [comment]
Yeah, I agree (regarding the need for a test that a developer can run). If the patch is in, then it must be something else. But I can't think of a reason to look up a keyspace name when doing a mutation write on a replica other than writing into a view.

@mykaul in any case this doesn't look as a regression or as having some material impact on users. It is still a serious bug in view code in my opinion, so I would not drop it off from the priority list completely. Best target IMO is 5.3. Re building a test case, I think it would be possible to build one now with test.py infrastructure.

**Comment by mykaul** [comment]
@kostja - 5.3 timeframe - still applicable here? Perhaps just to the develop the test mentioned above?

**Comment by mykaul** [comment]
> @kostja - 5.3 timeframe - still applicable here? Perhaps just to the develop the test mentioned above?

@kostja - same question - 5.4?

**Comment by kostja** [comment]
No.

**Comment by mykaul** [comment]
Moving to 6.1 for now.

**Comment by kbr-scylla** [comment]
@kostja from the above discussion it seems that the bug is in materialized views code, not in topology or schema code? In which case it should probably be Backend team instead of Cluster team.

I see it's already in Core - Backend backlog so unassigning from myself for now, if you decide it should be my team then reassign back.
